File: benchmarks/mse_sweep.py
#!/usr/bin/env python3
"""benchmarks/mse_sweep.py

Measures direct-lighting quality (MSE vs. a high-spp reference) for three
camera.h configurations - baseline (pre-RIS uniform-pick NEE), reservoir
(RIS only), and spatial (RIS + spatial reuse) - across scenes with
different light counts and at different samples-per-pixel budgets. Produces:

  benchmarks/results/mse_raw.csv       every individual run (one per repeat)
  benchmarks/results/mse_summary.csv   mean/std MSE per (light_count, spp, config)
  benchmarks/charts/mse_vs_spp.png     one panel per light count, MSE vs. spp
  benchmarks/charts/mse_vs_lights.png  one panel per spp, MSE vs. light count
  benchmarks/MSE_RESULTS.md            a short written summary

For each light count, one reference image is rendered once (spatial reuse
forced off - see mse_sweep.cpp) and reused for every spp/config/repeat at
that light count, so what's being measured is purely the technique and
sample count, not different ground truths. Test renders reuse the same
scene seed as their light count's reference (so geometry matches pixel for
pixel) but pick up independent Monte Carlo noise each repeat, since
camera.h's per-thread RNG is seeded from system entropy, not from that seed
- see mse_sweep.cpp's header comment.

Usage:
    python3 benchmarks/mse_sweep.py                  # full sweep - can take a while
    python3 benchmarks/mse_sweep.py --quick           # fast smoke test
    python3 benchmarks/mse_sweep.py --light-counts 1,4,12,32 --spp 4,16,64,128
    python3 benchmarks/mse_sweep.py --binary ../build/mse_sweep --repeats 5
"""
import argparse
import os
import pathlib
import statistics
import subprocess
import sys

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import pandas as pd

# Three configs compared at equal spp, at every (light_count, spp) point -
# see mse_sweep.cpp's header comment for exactly what each one does.
#   baseline   light_candidates=1, spatial_neighbors=0 - pre-RIS behavior
#   reservoir  light_candidates=N, spatial_neighbors=0 - RIS only
#   spatial    light_candidates=N, spatial_neighbors=K - RIS + spatial reuse
# Colors follow this project's existing amber/blue pair (colorblind-validated
# - see sweep.py) for baseline/reservoir, plus a third, distinguishable green
# for spatial; markers also differ per line so the distinction doesn't rely
# on color alone.
CONFIGS = {
    "baseline": dict(color="#c06a2c", marker="o", label="baseline (1 candidate, no spatial reuse)"),
    "reservoir": dict(color="#2f6fb5", marker="s", label="reservoir (RIS)"),
    "spatial": dict(color="#3c9a5f", marker="^", label="reservoir + spatial reuse"),
}
GRID_COLOR = "#dddddd"

CSV_COLUMNS = [
    "light_count", "spp", "config", "candidates", "spatial_neighbors",
    "spatial_radius", "seed", "mse", "seconds",
]


def scene_seed(seed_base, light_count):
    # One fixed seed per light count, shared by the reference and every test
    # render at that light count, so they all render the identical scene
    # geometry - the only thing allowed to differ is the lighting technique.
    return seed_base + light_count


def render_reference(binary, results_dir, cfg, light_count):
    ppm_path = results_dir / "ref_cache" / f"ref_L{light_count}_w{cfg['width']}_spp{cfg['ref_spp']}.ppm"
    ppm_path.parent.mkdir(parents=True, exist_ok=True)
    if ppm_path.exists():
        return ppm_path
    args = [
        str(binary), "--ref", str(cfg["width"]), str(cfg["depth"]), str(cfg["ref_spp"]),
        str(cfg["ref_candidates"]), str(cfg["threads"]), str(scene_seed(cfg["seed_base"], light_count)),
        str(light_count), str(ppm_path),
    ]
    result = subprocess.run(args, capture_output=True, text=True, check=True)
    print(f"  {result.stderr.strip()}")
    return ppm_path


def render_test(binary, cfg, light_count, spp, config_name, ref_ppm):
    spec = cfg["config_specs"][config_name]
    args = [
        str(binary), "--test", str(cfg["width"]), str(cfg["depth"]), str(spp),
        str(spec["candidates"]), str(spec["spatial_neighbors"]), str(cfg["spatial_radius"]),
        str(cfg["threads"]), str(scene_seed(cfg["seed_base"], light_count)), str(light_count),
        config_name, str(ref_ppm),
    ]
    result = subprocess.run(args, capture_output=True, text=True, check=True)
    last_line = result.stdout.strip().splitlines()[-1]
    return dict(zip(CSV_COLUMNS, last_line.split(",")))


def run_sweep(binary, results_dir, cfg, light_counts, spp_values, repeats):
    rows = []
    for light_count in light_counts:
        print(f"light_count={light_count}: rendering reference ({cfg['ref_spp']} spp)...")
        ref_ppm = render_reference(binary, results_dir, cfg, light_count)

        for spp in spp_values:
            for config_name in CONFIGS:
                mses = []
                for rep in range(repeats):
                    row = render_test(binary, cfg, light_count, spp, config_name, ref_ppm)
                    row["repeat"] = rep
                    rows.append(row)
                    mses.append(float(row["mse"]))
                print(f"  spp={spp:<5d} {config_name:<10s} mse mean={statistics.mean(mses):8.4f}"
                      + (f"  std={statistics.stdev(mses):.4f}" if repeats > 1 else "") + f"  (n={repeats})")
    return rows


def style_axes(ax):
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.spines["left"].set_color("#888888")
    ax.spines["bottom"].set_color("#888888")
    ax.grid(True, which="both", color=GRID_COLOR, linewidth=0.7, zorder=0)
    ax.set_axisbelow(True)
    ax.tick_params(colors="#444444", labelsize=8.5)


def plot_mse_vs_spp(df, light_counts, spp_values, out_path):
    n = len(light_counts)
    fig, axes = plt.subplots(1, n, figsize=(4.2 * n, 4), sharey=False)
    if n == 1:
        axes = [axes]
    fig.suptitle("Direct-lighting MSE vs. reference, by samples/pixel", fontsize=13,
                 fontweight="bold", color="#14181d")

    for ax, light_count in zip(axes, light_counts):
        sub = df[df["light_count"] == light_count]
        for config_name, style in CONFIGS.items():
            csub = sub[sub["config"] == config_name]
            agg = csub.groupby("spp")["mse"].agg(["mean", "std"]).reindex(spp_values)
            ax.errorbar(spp_values, agg["mean"], yerr=agg["std"].fillna(0), color=style["color"],
                        marker=style["marker"], markersize=5.5, linewidth=1.8, capsize=3,
                        label=style["label"])
        ax.set_xscale("log", base=2)
        ax.set_yscale("log")
        ax.set_xticks(spp_values)
        ax.set_xticklabels([str(v) for v in spp_values])
        ax.set_title(f"{light_count} light{'s' if light_count != 1 else ''}", fontsize=10, color="#5b6570")
        ax.set_xlabel("samples/pixel (log scale)", fontsize=8.5, color="#5b6570")
        if ax is axes[0]:
            ax.set_ylabel("MSE vs. reference (log scale)", fontsize=8.5, color="#5b6570")
        style_axes(ax)

    axes[-1].legend(frameon=False, fontsize=8, loc="upper right")
    fig.tight_layout(rect=(0, 0, 1, 0.92))
    fig.savefig(out_path, dpi=150, facecolor="white")
    plt.close(fig)


def plot_mse_vs_lights(df, light_counts, spp_values, out_path):
    n = len(spp_values)
    fig, axes = plt.subplots(1, n, figsize=(4.2 * n, 4), sharey=False)
    if n == 1:
        axes = [axes]
    fig.suptitle("Direct-lighting MSE vs. reference, by light count", fontsize=13,
                 fontweight="bold", color="#14181d")

    for ax, spp in zip(axes, spp_values):
        sub = df[df["spp"] == spp]
        for config_name, style in CONFIGS.items():
            csub = sub[sub["config"] == config_name]
            agg = csub.groupby("light_count")["mse"].agg(["mean", "std"]).reindex(light_counts)
            ax.errorbar(light_counts, agg["mean"], yerr=agg["std"].fillna(0), color=style["color"],
                        marker=style["marker"], markersize=5.5, linewidth=1.8, capsize=3,
                        label=style["label"])
        ax.set_yscale("log")
        ax.set_xticks(light_counts)
        ax.set_title(f"{spp} spp", fontsize=10, color="#5b6570")
        ax.set_xlabel("lights in scene", fontsize=8.5, color="#5b6570")
        if ax is axes[0]:
            ax.set_ylabel("MSE vs. reference (log scale)", fontsize=8.5, color="#5b6570")
        style_axes(ax)

    axes[-1].legend(frameon=False, fontsize=8, loc="upper right")
    fig.tight_layout(rect=(0, 0, 1, 0.92))
    fig.savefig(out_path, dpi=150, facecolor="white")
    plt.close(fig)


def write_results_md(summary, light_counts, spp_values, repeats, out_path, chart1, chart2):
    lines = [
        "# Direct-lighting MSE sweep results",
        "",
        f"Generated by `benchmarks/mse_sweep.py` ({repeats} repeat(s) per configuration; "
        "each repeat is an independent Monte Carlo noise draw on identical scene geometry - "
        "see the script's docstring). Lower MSE is better (closer to the high-spp reference).",
        "",
        f"![MSE vs. samples per pixel]({chart1.name})",
        "",
        f"![MSE vs. light count]({chart2.name})",
        "",
        "## Mean MSE by configuration",
        "",
        "| lights | spp | baseline | reservoir | +spatial | reservoir vs. baseline | +spatial vs. reservoir |",
        "|---|---|---|---|---|---|---|",
    ]
    for light_count in light_counts:
        for spp in spp_values:
            row = summary[(summary["light_count"] == light_count) & (summary["spp"] == spp)]
            vals = {name: row[row["config"] == name]["mse_mean"].iloc[0] for name in CONFIGS
                    if not row[row["config"] == name].empty}
            if len(vals) < 3:
                continue
            base, res, spa = vals["baseline"], vals["reservoir"], vals["spatial"]
            res_ratio = base / res if res > 0 else float("inf")
            spa_ratio = res / spa if spa > 0 else float("inf")
            lines.append(f"| {light_count} | {spp} | {base:.4f} | {res:.4f} | {spa:.4f} "
                         f"| {res_ratio:.2f}x lower | {spa_ratio:.2f}x lower |")
    lines.append("")
    out_path.write_text("\n".join(lines))


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--binary", default=None, help="path to the mse_sweep binary (default: ../build/mse_sweep)")
    parser.add_argument("--light-counts", default="1,4,12,32", help="comma-separated light counts to sweep")
    parser.add_argument("--spp", default="4,16,64,128", help="comma-separated samples-per-pixel budgets to sweep")
    parser.add_argument("--width", type=int, default=200)
    parser.add_argument("--depth", type=int, default=8)
    parser.add_argument("--ref-spp", type=int, default=512)
    parser.add_argument("--ref-candidates", type=int, default=8)
    parser.add_argument("--candidates", type=int, default=4, help="light_candidates for the reservoir/spatial configs")
    parser.add_argument("--spatial-neighbors", type=int, default=4)
    parser.add_argument("--spatial-radius", type=int, default=20)
    parser.add_argument("--threads", type=int, default=0, help="0 = let the binary pick (hardware_concurrency)")
    parser.add_argument("--repeats", type=int, default=3, help="independent renders per (light_count, spp, config)")
    parser.add_argument("--seed-base", type=int, default=1000)
    parser.add_argument("--quick", action="store_true", help="fast smoke test: fewer values, 1 repeat, tiny image")
    args = parser.parse_args()

    here = pathlib.Path(__file__).resolve().parent
    binary = pathlib.Path(args.binary) if args.binary else here.parent / "build" / "mse_sweep"
    if not binary.exists():
        sys.exit(f"mse_sweep binary not found at {binary}\n"
                  f"Build it first: mkdir -p build && cd build && cmake .. -DCMAKE_BUILD_TYPE=Release && cmake --build .")

    light_counts = [int(v) for v in args.light_counts.split(",")]
    spp_values = [int(v) for v in args.spp.split(",")]
    repeats = args.repeats
    width, depth, ref_spp = args.width, args.depth, args.ref_spp

    if args.quick:
        light_counts = light_counts[:2] if len(light_counts) > 2 else light_counts
        spp_values = spp_values[:2] if len(spp_values) > 2 else spp_values
        repeats = 1
        width = min(width, 120)
        depth = min(depth, 6)
        ref_spp = min(ref_spp, 96)

    threads = args.threads if args.threads > 0 else (os.cpu_count() or 4)

    cfg = dict(
        width=width, depth=depth, ref_spp=ref_spp, ref_candidates=args.ref_candidates,
        threads=threads, seed_base=args.seed_base, spatial_radius=args.spatial_radius,
        config_specs={
            "baseline": dict(candidates=1, spatial_neighbors=0),
            "reservoir": dict(candidates=args.candidates, spatial_neighbors=0),
            "spatial": dict(candidates=args.candidates, spatial_neighbors=args.spatial_neighbors),
        },
    )

    results_dir = here / "results"
    charts_dir = here / "charts"
    for d in (results_dir, charts_dir):
        d.mkdir(parents=True, exist_ok=True)

    print(f"Using binary: {binary}")
    print(f"light_counts={light_counts}  spp={spp_values}  repeats={repeats}  "
          f"width={width}  ref_spp={ref_spp}  threads={threads}\n")

    rows = run_sweep(binary, results_dir, cfg, light_counts, spp_values, repeats)

    df = pd.DataFrame(rows)
    for col in ["light_count", "spp", "candidates", "spatial_neighbors", "spatial_radius", "seed", "repeat"]:
        df[col] = df[col].astype(int)
    for col in ["mse", "seconds"]:
        df[col] = df[col].astype(float)
    raw_path = results_dir / "mse_raw.csv"
    df.to_csv(raw_path, index=False)
    print(f"\nWrote {len(df)} raw rows -> {raw_path}")

    summary = df.groupby(["light_count", "spp", "config"]).agg(
        mse_mean=("mse", "mean"), mse_std=("mse", "std"),
        seconds_mean=("seconds", "mean"),
    ).reset_index()
    summary_path = results_dir / "mse_summary.csv"
    summary.to_csv(summary_path, index=False)
    print(f"Wrote {summary_path}")

    chart1 = charts_dir / "mse_vs_spp.png"
    plot_mse_vs_spp(df, light_counts, spp_values, chart1)
    print(f"Wrote {chart1}")

    chart2 = charts_dir / "mse_vs_lights.png"
    plot_mse_vs_lights(df, light_counts, spp_values, chart2)
    print(f"Wrote {chart2}")

    results_md = here / "MSE_RESULTS.md"
    write_results_md(summary, light_counts, spp_values, repeats, results_md, chart1, chart2)
    print(f"Wrote {results_md}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: benchmarks/mse_sweep.cpp
// benchmarks/mse_sweep.cpp - measures direct-lighting quality (MSE vs. a
// high-spp reference) for three camera.h configurations, across scenes with
// different light counts and at different samples-per-pixel budgets:
//
//   baseline   light_candidates=1, spatial_neighbors=0
//              mathematically identical to plain uniform-pick NEE, before
//              RIS/reservoir sampling existed - see camera.h's header comment.
//   reservoir  light_candidates=N (default 4), spatial_neighbors=0
//              RIS only: each shading point scores N candidates and keeps
//              the best one, still one shadow ray per hit.
//   spatial    light_candidates=N, spatial_neighbors=K (default 4)
//              RIS + spatial reuse: the primary vertex additionally folds
//              in K neighboring pixels' already-built reservoirs before
//              resolving - see spatially_combined_reservoir() in camera.h.
//
// Scenes: same "random spheres + a ring of small lights" layout as the
// project's default scene (src/main.cpp), but the ring's light count is the
// swept axis - more/smaller lights is exactly the case uniform-pick NEE
// gets noisiest on and RIS/spatial reuse are meant to help with.
//
// This binary renders ONE data point per invocation - the reference
// (--ref) or one test config (--test) - and is driven repeatedly by
// mse_sweep.py, which handles the light-count/spp/repeats loops, caches
// the reference per light count, and turns the resulting CSV into charts.
// Same split as main.cpp's --bench/--scene modes, for the same reason:
// keep the C++ side a dumb, deterministic single-shot renderer and put
// orchestration/statistics/plotting in Python.
#include <chrono>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <random>
#include <string>
#include <thread>
#include <vector>

#include "../src/bvh.h"
#include "../src/camera.h"
#include "../src/hittable_list.h"
#include "../src/material.h"
#include "../src/sphere.h"
#include "../src/vec3.h"

// A "many small lights" scene: a field of diffuse/metal/glass spheres (same
// spirit as the project's main scene) plus a ring of `light_count` small
// emissive spheres. `light_count` is the axis this harness sweeps - bigger
// rings mean a bigger pool for uniform-pick NEE to search blindly, which is
// exactly where RIS/reservoir sampling (and then spatial reuse on top)
// should show a growing advantage.
hittable_list build_lit_scene(std::vector<std::shared_ptr<sphere>>& lights_out, unsigned int seed,
                               int light_count) {
    hittable_list world;
    lights_out.clear();
    std::mt19937 rng(seed);

    auto ground = std::make_shared<lambertian>(color(0.5, 0.5, 0.5));
    world.add(std::make_shared<sphere>(point3(0, -1000, 0), 1000, ground));

    for (int a = -6; a < 6; a++) {
        for (int b = -6; b < 6; b++) {
            double choose = random_double(rng);
            point3 center(a + 0.9 * random_double(rng), 0.2, b + 0.9 * random_double(rng));
            if ((center - point3(4, 0.2, 0)).length() > 0.9) {
                std::shared_ptr<material> mat;
                if (choose < 0.7) {
                    mat = std::make_shared<lambertian>(color::random(rng) * color::random(rng));
                } else if (choose < 0.9) {
                    mat = std::make_shared<metal>(color::random(rng, 0.5, 1), random_double(rng, 0, 0.4));
                } else {
                    mat = std::make_shared<dielectric>(1.5);
                }
                world.add(std::make_shared<sphere>(center, 0.2, mat));
            }
        }
    }

    for (int i = 0; i < light_count; i++) {
        double angle = (2.0 * pi * i) / std::max(1, light_count);
        double r = 6.0;
        point3 pos(r * std::cos(angle), 2.5 + 1.5 * ((i % 3) - 1), r * std::sin(angle));
        color c = color::random(rng, 3.0, 7.0);
        auto light_mat = std::make_shared<diffuse_light>(c);
        auto light = std::make_shared<sphere>(pos, 0.3, light_mat);
        world.add(light);
        lights_out.push_back(light);
    }

    return hittable_list(std::make_shared<bvh_node>(world));
}

camera make_camera(int width, int spp, int depth, int candidates, int spatial_neighbors, int spatial_radius) {
    camera cam;
    cam.aspect_ratio = 1.5;
    cam.image_width = width;
    cam.samples_per_pixel = spp;
    cam.max_depth = depth;
    cam.light_candidates = candidates;
    cam.spatial_neighbors = spatial_neighbors;
    cam.spatial_radius = spatial_radius;
    cam.vfov = 40;
    cam.lookfrom = point3(0, 3, 9);
    cam.lookat = point3(0, 1, 0);
    cam.vup = vec3(0, 1, 0);
    cam.defocus_angle = 0;
    return cam;
}

double mse_vs_reference(const std::vector<uint8_t>& test, const std::vector<uint8_t>& ref) {
    double sum = 0;
    for (size_t i = 0; i < test.size(); i++) {
        double d = static_cast<double>(test[i]) - static_cast<double>(ref[i]);
        sum += d * d;
    }
    return sum / test.size();
}

void write_ppm(const std::string& path, int w, int h, const std::vector<uint8_t>& px) {
    FILE* f = std::fopen(path.c_str(), "wb");
    if (!f) {
        std::fprintf(stderr, "error: cannot write %s\n", path.c_str());
        std::exit(1);
    }
    std::fprintf(f, "P6\n%d %d\n255\n", w, h);
    std::fwrite(px.data(), 1, px.size(), f);
    std::fclose(f);
}

std::vector<uint8_t> read_ppm(const std::string& path, int& w, int& h) {
    FILE* f = std::fopen(path.c_str(), "rb");
    if (!f) {
        std::fprintf(stderr, "error: cannot read %s (render the reference first with --ref)\n", path.c_str());
        std::exit(1);
    }
    char magic[3] = {0};
    int maxval = 0;
    if (std::fscanf(f, "%2s %d %d %d", magic, &w, &h, &maxval) != 4 || std::strcmp(magic, "P6") != 0) {
        std::fprintf(stderr, "error: %s is not a well-formed P6 PPM\n", path.c_str());
        std::exit(1);
    }
    std::fgetc(f);  // single whitespace byte separating the header from pixel data
    std::vector<uint8_t> px(static_cast<size_t>(w) * h * 3);
    size_t n = std::fread(px.data(), 1, px.size(), f);
    std::fclose(f);
    if (n != px.size()) {
        std::fprintf(stderr, "error: %s is truncated\n", path.c_str());
        std::exit(1);
    }
    return px;
}

void print_usage(const char* prog) {
    std::fprintf(stderr,
                  "Usage:\n"
                  "  %s --ref  <width> <depth> <ref_spp> <ref_candidates> <threads> <seed> "
                  "<light_count> <out_ppm>\n"
                  "  %s --test <width> <depth> <spp> <candidates> <spatial_neighbors> "
                  "<spatial_radius> <threads> <seed> <light_count> <config_label> <ref_ppm>\n"
                  "\n"
                  "--ref renders a high-spp ground truth (spatial reuse forced off - see the\n"
                  "comment at its call site) and saves it as a PPM. --test renders one config,\n"
                  "compares it against that PPM by mean squared error, and prints one CSV line\n"
                  "to stdout:\n"
                  "  light_count,spp,config,candidates,spatial_neighbors,spatial_radius,seed,mse,seconds\n"
                  "\n"
                  "Normally driven by mse_sweep.py - see benchmarks/README.md.\n",
                  prog, prog);
}

int main(int argc, char** argv) {
    if (argc < 2) {
        print_usage(argv[0]);
        return 1;
    }
    std::string mode = argv[1];

    if (mode == "--ref") {
        if (argc < 10) {
            print_usage(argv[0]);
            return 1;
        }
        int width = std::atoi(argv[2]);
        int depth = std::atoi(argv[3]);
        int ref_spp = std::atoi(argv[4]);
        int ref_candidates = std::atoi(argv[5]);
        int threads = std::atoi(argv[6]);
        unsigned int seed = static_cast<unsigned int>(std::strtoul(argv[7], nullptr, 10));
        int light_count = std::atoi(argv[8]);
        std::string out_ppm = argv[9];

        std::vector<std::shared_ptr<sphere>> lights;
        hittable_list world = build_lit_scene(lights, seed, light_count);

        // Spatial reuse stays off for the reference: its Jacobian-skip is a
        // documented, deliberate small-bias simplification (see the "Add
        // spatial reuse..." commit and spatially_combined_reservoir()'s
        // comment in camera.h) - ground truth shouldn't inherit that bias.
        // A generous candidate count is fine here since plain RIS is
        // unbiased regardless of width; it just converges faster.
        camera cam = make_camera(width, ref_spp, depth, ref_candidates, /*spatial_neighbors=*/0,
                                  /*spatial_radius=*/0);

        auto t0 = std::chrono::high_resolution_clock::now();
        std::vector<uint8_t> pixels = cam.render(world, lights, threads);
        auto t1 = std::chrono::high_resolution_clock::now();
        std::fprintf(stderr, "reference: light_count=%d width=%d ref_spp=%d -> %s (%.1fs)\n", light_count,
                      width, ref_spp, out_ppm.c_str(), std::chrono::duration<double>(t1 - t0).count());
        write_ppm(out_ppm, width, cam.height(), pixels);
        return 0;
    }

    if (mode == "--test") {
        if (argc < 13) {
            print_usage(argv[0]);
            return 1;
        }
        int width = std::atoi(argv[2]);
        int depth = std::atoi(argv[3]);
        int spp = std::atoi(argv[4]);
        int candidates = std::atoi(argv[5]);
        int spatial_neighbors = std::atoi(argv[6]);
        int spatial_radius = std::atoi(argv[7]);
        int threads = std::atoi(argv[8]);
        unsigned int seed = static_cast<unsigned int>(std::strtoul(argv[9], nullptr, 10));
        int light_count = std::atoi(argv[10]);
        std::string config_label = argv[11];
        std::string ref_ppm = argv[12];

        std::vector<std::shared_ptr<sphere>> lights;
        // Same seed as the --ref call that built ref_ppm -> identical scene
        // geometry, so the only thing differing between test and reference
        // is the direct-lighting technique and sample count being measured.
        hittable_list world = build_lit_scene(lights, seed, light_count);
        camera cam = make_camera(width, spp, depth, candidates, spatial_neighbors, spatial_radius);

        auto t0 = std::chrono::high_resolution_clock::now();
        std::vector<uint8_t> pixels = cam.render(world, lights, threads);
        auto t1 = std::chrono::high_resolution_clock::now();
        double seconds = std::chrono::duration<double>(t1 - t0).count();

        int ref_w = 0, ref_h = 0;
        std::vector<uint8_t> ref_pixels = read_ppm(ref_ppm, ref_w, ref_h);
        if (ref_w != width || ref_h != cam.height()) {
            std::fprintf(stderr, "error: reference %s is %dx%d, test render is %dx%d - can't compare\n",
                          ref_ppm.c_str(), ref_w, ref_h, width, cam.height());
            return 1;
        }

        double mse = mse_vs_reference(pixels, ref_pixels);
        std::fprintf(stderr, "  light_count=%-3d spp=%-4d %-10s candidates=%d spatial_neighbors=%d -> mse=%.4f (%.1fs)\n",
                      light_count, spp, config_label.c_str(), candidates, spatial_neighbors, mse, seconds);
        std::printf("%d,%d,%s,%d,%d,%d,%u,%.6f,%.3f\n", light_count, spp, config_label.c_str(), candidates,
                     spatial_neighbors, spatial_radius, seed, mse, seconds);
        return 0;
    }

    print_usage(argv[0]);
    return 1;
}
